tensorboard/writer/record_writer.py

```diff
@@ -0,0 +1,158 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""
+To write tf_record into file. Here we use it for tensorboard's event writting.
+The code was borrowed from https://github.com/TeamHG-Memex/tensorboard_logger
+"""
+
+import copy
+import io
+import os.path
+import re
+import struct
+try:
+    import boto3
+    S3_ENABLED = True
+except ImportError:
+    S3_ENABLED = False
+
+from .crc32c import crc32c
+
+
+_VALID_OP_NAME_START = re.compile('^[A-Za-z0-9.]')
+_VALID_OP_NAME_PART = re.compile('[A-Za-z0-9_.\\-/]+')
+
+# Registry of writer factories by prefix backends.
+#
+# Currently supports "s3://" URLs for S3 based on boto and falls
+# back to local filesystem.
+REGISTERED_FACTORIES = {}
+
+
+def register_writer_factory(prefix, factory):
+    if ':' in prefix:
+        raise ValueError('prefix cannot contain a :')
+    REGISTERED_FACTORIES[prefix] = factory
+
+
+def directory_check(path):
+    '''Initialize the directory for log files.'''
+    try:
+        prefix = path.split(':')[0]
+        factory = REGISTERED_FACTORIES[prefix]
+        return factory.directory_check(path)
+    except KeyError:
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+
+def open_file(path):
+    '''Open a writer for outputting event files.'''
+    try:
+        prefix = path.split(':')[0]
+        factory = REGISTERED_FACTORIES[prefix]
+        return factory.open(path)
+    except KeyError:
+        return open(path, 'wb')
+
+
+class S3RecordWriter(object):
+    """Writes tensorboard protocol buffer files to S3."""
+
+    def __init__(self, path):
+        if not S3_ENABLED:
+            raise ImportError("boto3 must be installed for S3 support.")
+        self.path = path
+        self.buffer = io.BytesIO()
+
+    def __del__(self):
+        self.close()
+
+    def bucket_and_path(self):
+        path = self.path
+        if path.startswith("s3://"):
+            path = path[len("s3://"):]
+        bp = path.split("/")
+        bucket = bp[0]
+        path = path[1 + len(bucket):]
+        return bucket, path
+
+    def write(self, val):
+        self.buffer.write(val)
+
+    def flush(self):
+        s3 = boto3.client('s3')
+        bucket, path = self.bucket_and_path()
+        upload_buffer = copy.copy(self.buffer)
+        upload_buffer.seek(0)
+        s3.upload_fileobj(upload_buffer, bucket, path)
+
+    def close(self):
+        self.flush()
+
+
+class S3RecordWriterFactory(object):
```

Comment:
Let's remove S3 support for now and we'll add that by extending gfile in the compat layer in a later PR.